incubator/hnc/internal/reconcilers/anchor.go

```diff
@@ -87,6 +87,14 @@ func (r *AnchorReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		return ctrl.Result{}, err
 	}
 
+	// In v0.5.x, HNC uses v1alpha1 and the annotation is called "subnamespaceOf".
+	// In v1alpha2, the namespace annotation is upgraded into "subnamespace-of".
+	// If the deprecated "subnamespaceOf" annotation still exists, upgrade API and
+	// early exit since updating the subnamespace will enqueue the anchor again.
+	if oldSOf := snsInst.GetAnnotations()[api.DeprecatedSubnamespaceOf]; oldSOf != "" {
+		return ctrl.Result{}, r.upgradeSubnamespaceOfAPI(ctx, log, snsInst, oldSOf)
+	}
+
```

Comment:
This feels unnatural to me, tbh. This reconciler has only ever created and deleted namespaces, not updated them - the HC reconciler has a much better flow for making incremental changes to namespaces. It would also allow you to leave most of this file completely untouched - you wouldn't need to change the mapping function or add a getSubnamespaceOf() function (when HNC v0.6 is installed, all anchors would briefly go into a Conflict state until their annotations were fixed, but that's fine IMO).

Reply:
sg will look into it.

Reply:
In your other PR changing the anchor reconciler flow, if the state is `Conflict`, it will be untouched and nothing will be deleted, right? If so, it should be fine to have `Conflict` state for a short period at HNC startup.

Reply:
Moved the change to HC reconciler and leave anchor reconciler untouched. All tests passed. Just need a confirmation for my above comment from @adrianludwin .

Reply:
Yes, that's right. If an anchor is in the "conflict" state, it just means that if you delete the anchor, the namespace won't be deleted. As soon as the annotation is fixed, that problem goes away. Note that this doesn't affect the _parent_ of the subnamespace, which remains unchanged throughout.